closure/goog/i18n/numberformat.js

```diff
@@ -624,22 +624,104 @@ goog.i18n.NumberFormat.prototype.subformatFixed_ =
+        if (digitLen - i > 1) {
+          currentGroupSize = this.groupingArray_[currentGroupSizeIndex];
+          if (i < repeatedDigitLen) {
+            // Process the left side (the repeated number groups)
+            var repeatedDigitIndex = repeatedDigitLen - i;
+            // Edge case if there's a number grouping asking for "1" group at
+            // a time; otherwise, if the remainder is 1, there's the separator
+            if (currentGroupSize === 1 ||
+                (currentGroupSize > 0 &&
+                (repeatedDigitIndex % currentGroupSize) === 1)
+            ) {
+              parts.push(grouping);
+            }
+          } else {
+            // Process the right side (the non-repeated fixed number groups)
+            if (currentGroupSizeIndex < this.groupingArray_.length) {
+              if (i === repeatedDigitLen) {
+                // Increase the group index because a separator
+                // has previously added in the earlier logic
+                currentGroupSizeIndex += 1;
+              } else {
+                // Otherwise, just iterate to the right side and
+                // add a separator once the length matches to the expected
+                if (currentGroupSize ===
+                    i - repeatedDigitLen - nonRepeatedGroupCompleteCount + 1) {
+                  parts.push(grouping);
+                  // Keep track of what has been completed on the right
+                  nonRepeatedGroupCompleteCount += currentGroupSize;
+                  currentGroupSizeIndex += 1; // Get to the next number grouping
+                }
+              }
+            }
+          }
+        }
       }
+    } else {
+      // There are no repeating digits and only non-repeating digits
+      var digitLenLeft = digitLen;
+      var rightToLeftParts = [];
+      // Start from the right most non-repeating group and work inwards
+      for (
+          currentGroupSizeIndex = this.groupingArray_.length - 1;
+          currentGroupSizeIndex >= 0 && digitLenLeft > 0;
+          currentGroupSizeIndex--) {
+        currentGroupSize = this.groupingArray_[currentGroupSizeIndex];
+        // Iterate from the right most digit
+        for (var rightDigitIndex = 0;
+             rightDigitIndex < currentGroupSize && (
+             (digitLenLeft - rightDigitIndex - 1) >= 0
+             );
+             rightDigitIndex++) {
+          rightToLeftParts.push(
+              String.fromCharCode(
+                  zeroCode + intPart.charAt(
+                      digitLenLeft - rightDigitIndex - 1
+                  ) * 1
+              )
+          );
+        }
+        // Update the number of digits left
+        digitLenLeft -= currentGroupSize;
+        if (digitLenLeft > 0) {
+          rightToLeftParts.push(grouping);
+        }
+      }
+      // Reverse and push onto the remaining parts
+      rightToLeftParts.reverse().forEach(function(item) {
+        parts.push(item);
+      });
```

Comment:
High level comment: you're adding a bunch of dense code to an already large method. I think it would greatly help readability/comprehension if you could break some of this logic up into discrete testable functions rather than a monolithic chunk of code.